### ISSUE TITLE
Object has been deleted or invalidated - only throwing error on device that deletes the object.

### ISSUE THREAD

**Issue by tlynchlaw** [realm/realm-cocoa#6376]
I am a fairly new developer and I am using Xcode and Swift 5.1.

I am getting the error above when I delete a Member object from the Realm on the device that deletes it.  All other devices running at the same time handle the deletion fine.  Here is the setup:

1. I have Members in a firestore database.
1. I have a firestore observer that observes changes in that database and writes the changes to Realm on the device.
1. In viewWillAppear in the MainVC I load the Members from Realm.  And I set up a notification listener to observe changes in the Members in Realm.
1. In a SettingsVC, I also also load the Members from Realm and set up a notification listener there as well.
1. In the SettingsVC, I delete a Member using a google cloud function, which removes the member from the firestore database.
1. My firestore observer picks up the changes and removes the Member from the Realm file.
1. The listener in the SettingsVC, where the removal was done, picks up the changes in Realm and removes the Member from the tableView.
1. Then it crashes with the error above.
1. Note, if I have 3 simulators open at the same time and running the same version of the app, only the simulator (or device) that actually deletes the Member crashes.  The other two remove the Member from the tableview and do not crash.

Here is the code found in the MainVC and the SettingsVC to load the Members and set up the notification listener:

```
var members:Results<Member>?
```

```
    // Get the members from Realm.
    UserService.loadMembers { (members) in
        self.members = members
    }
    
    // Set up a listener for changes in the members.
    memberNotificationToken = members?.observe { [weak self] (changes: RealmCollectionChange) in
        // Make sure tableView has not been dismissed
        guard let tableView = self?.tableView else { return }
        tableView.reloadData()
    }
}

// Turn off the listener.
deinit {
    memberNotificationToken?.invalidate()
}
```

Here is the code in UserService.loadMembers:

```
static func loadMembers(completion: @escaping (_ members:Results<Member>?) -> Void) {
    
    // Get a reference to the Realm database where the members are stored.
    let realm = try! Realm()
    // Get the members.
    let members = realm.objects(Member.self).sorted(byKeyPath: "displayName")
    // Send them back.
    completion (members)
}
```

Here is the code to delete the Member:

```
            UserService.deleteUser(memberToDelete: memberToDelete, completion: { status in
                if status != nil {
                    Alerts.showBasicAlert(title: Generic Title, message: Generic message) { (alertController) in
                        self.present(alertController, animated: true)
                    }
                    // Reload the table.
                    tableView.reloadData()
                }
            })
        })
```

## Version of Realm and Tooling
Realm 4.1.0
Xcode version: 
11.2.1
iOS/OSX version: ?
13.0

Any help would be appreciated.  I've been trying to solve this for two days.

**Comment by ppamorim** [comment]
Are you sure that this code below is not async?

```swift
UserService.loadMembers { (members) in
    self.members = members
}
```

**Comment by tlynchlaw** [comment]
It likely is. How would I fix?

**Comment by tlynchlaw** [comment]
It crashed.  I made this change to the SettingsVC:

        // Get the members from Realm.
        UserService.loadMembers { [weak self] (members) in
            self!.members = members
            
            // Set up a listener for changes in the members.
            self!.memberNotificationToken = members?.observe { [weak self] (changes: RealmCollectionChange) in
                // Make sure tableView has not been dismissed
                guard let tableView = self?.tableView else { return }
                tableView.reloadData()
            }
        }

I also tried it without the [weak self] on the first line.  I did not make the same change to the MainVC.  Will try that now.

**Comment by tlynchlaw** [comment]
When I made the same changes to the MainVC it worked the first time with no crash.  But then I deleted another member and it crashed.

**Comment by ppamorim** [comment]
When using `[weak self]`, use `self?.members` for example. This is safer.
Could you please config if `memberToDelete` is a managed object?
Also remember that deleting a Realm object will make the content of `Results<Member>` have items that are invalidated if you are keeping a strong reference of them.

**Comment by ppamorim** [comment]
If crashing because you are trying to read an managed object that has been deleted. You can use the var `isInvalidated` to verify if the managed object is valid or not.

**Comment by tlynchlaw** [comment]
I believe it is a managed object, but I am not familiar with what makes it managed or not.  Here is the code:        
        // Get a reference to the member to be deleted
        let memberToDelete = members![indexPath.row]

I read about using `isInvalidated`, but I also read it was cumbersome and itself had to be managed.  I also don't know how to use it.

I just ran the code again and stepped through my tableview code.  I printed out the members and indeed, they were correctly reduced from 3 to 2.  But when the tableview went to do the index of the one that was deleted (which happened to be the first one), it crashed.  So I see what you are saying, but I don't know how to fix it.

**Comment by ppamorim** [comment]
Could you please share the code of the tableview?

**Comment by tlynchlaw** [comment]
Here's cellForRowAt:

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        //Set up a cell
        let cell = tableView.dequeueReusableCell(withIdentifier: Constants.Storyboard.userCell, for: indexPath) as! UserCell
        // Get the Member info for this row
        let member = members?[indexPath.row]
        // Set up the labels
        guard member != nil else { return cell }
        cell.setMember(member!)
        
        return cell
    }

**Comment by ppamorim** [comment]
When you call `cell.setMember(member!)`, is it simply reading `member` or is it storing member inside the cell?

misc: You can rewrite it to:

```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    //Set up a cell
    let cell = tableView.dequeueReusableCell(withIdentifier: Constants.Storyboard.userCell, for: indexPath)

    // Get the Member info for this row
    guard let member = members?[indexPath.row] else {
         return cell
    }
    // Set up the labels
    (cell as? UserCell)?.setMember(member)
    
    return cell
}
```

**Comment by tlynchlaw** [comment]
It sends `member` into the custom cell function setMember, where its display name and other characteristics are read and used to populate the labels in the cell.

Thanks for the other code, too.  I will study it to learn how to code better.

**Comment by tlynchlaw** [comment]
Here is setMember:

    func setMember(_ member:Member) {
            
        // Put the member's name in the cell.  Get the display name first.
        UserService.getDisplayName(member.userId) { (name) in
            self.nameLabel.text = name
        }
        // Show or hide pending, depending ... ha ha.
        if member.pending {
            pendingLabel.alpha = 1
        } else {
            pendingLabel.alpha = 0
        }
    }
}

**Comment by ppamorim** [comment]
as debug, try to add this and add a breakpoint in the return:

```
func setMember(_ member:Member) {

    if member.isInvalidated {
        return
    }
        
    // Put the member's name in the cell.  Get the display name first.
    UserService.getDisplayName(member.userId) { (name) in
```

**Comment by tlynchlaw** [comment]
I added the code and put a breakpoint on `return`.  It crashed without breaking at that point.

**Comment by tlynchlaw** [comment]
FYI, this is `getDisplayName` in the `UserService` also:

    static func getDisplayName(_ userId:String, completion: @escaping (_ displayName:String) -> Void) {
        
        let realm = try! Realm()
        // Get the display name from the post.
        let member = realm.objects(Member.self).filter("userId = '\(userId)'").first
        let displayName = member?.displayName ?? "Unknown"
        completion (displayName)
    }

**Comment by ppamorim** [comment]
That's expected. Your `Member` object is getting invalidated upon reading but that check is prevening the use of the deleted member instance.

**Comment by ppamorim** [comment]
Try to temporarly remove this `getDisplayName` function and see if it crashes.

**Comment by tlynchlaw** [comment]
same crash...

**Comment by ppamorim** [comment]
So the problem is that the `Member` object is getting a invalidated object. The `Results` list probably is not updated when you call the tableview function so it's getting a old version of the `Results`.
I would recommend you to move the data load to the function `func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath)` and keep the `func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)` only for the `tableView.dequeueReusableCell`. 

Also make sure to only reload the UITableView when your observer detects a data change.

**Comment by tlynchlaw** [comment]
I have not worked with `willDisplay cell.`  I will review it and see if I can get it to work.  Thank you.

**Comment by tlynchlaw** [comment]
@ppamorim I am thinking the answer must be fairly obvious and I'm just missing it.  I use the exact same approach with the comments on the MainVC.  The only difference is that I do everything on that VC (delete and reload) and there is no other VC involved.  Realm says to ask on StackOverflow, so I guess I will try there.  Thanks for all your input and help.

**Comment by tlynchlaw** [comment]
@ppamorim   I fixed the problem. It was a stupid mistake as so many of my questions are. What happened was: in my alert, which shows after successful deletion, I was attempting to show the name of the member that was deleted. But of course that member was already gone. I found it by using a symbolic breakpoint with [RLMException]. That took me to the line, and then I worked backwards. That was a first for me using this debugging technique. Thank you for taking the time to respond to my question. I very much appreciate all your time.
